Created abstract component loader class

alexandrevryghem opened the pull request:
## References
* Related to #2235

## Description
While implementing PR #2235, I realised that there were a multiple `ComponentLoader`s which shared the same code like `ngOnChanges()`, `instantiateComponent()` & `connectInputsAndOutputs()`. Creating an abstract parent class for all the component loaders will make it easier to fix bugs/add features in the future.

## Instructions for Reviewers
List of changes in this PR:
- Created an `AbstractComponentLoaderComponent` and made `AdvancedWorkflowActionsLoaderComponent`, `ClaimedTaskActionsLoaderComponent`, `ListableObjectComponentLoaderComponent`, `MetadataRepresentationLoaderComponent` & `BrowseBySwitcherComponent` extend them and moved common part to the abstract class
- Used the same html template for all the component loaders and created a `DynamicComponentLoaderDirective` that replaces all the individual component loader directives (like `AdvancedWorkflowActionsDirective`, `ListableObjectDirective`, ...)
- During PR #2235 the `ngOnChanges` fix was not applied to `AdvancedWorkflowActionsLoaderComponent` yet so this is now also fixed
- Fixed issue where the matched components was not recalculated when one of its `@Input()`s he uses to find a match is updated. To keep track of this a new list has been created `inputNamesDependentForComponent`
- Separated the `inAndOutputNames` list into two separate list `inputNames` & `outputNames`, in order to be able to use Angular's new `seInput()` method to update its `@Input()`s. The advantage of this is that we now don't have to manually call the `ngOnChanges` anymore because it will automatically be called. (It also has some benefits for the tests, previously when you used `await fixture.whenStable()` it wasn't able to detect if the fixture was also stabilised in the child components when they were created dynamically. Now Angular will be able to wait for those changes as well). I will also open a PR to implement this for the normal ThemedComponents when I have time
- Fixed bug in `BrowseService#getBrowseEntriesFor` & `BrowseService#getBrowseItemsFor` where the `useCachedVersionIfAvailable` & `requestOnStale` where set to `null` instead of `undefined`. This meant that the value `null` was used instead of the default values.

**Guidance for how to test or review this PR:**
- Go to pages with usages for the following components and check that they all still work like before
  - `AdvancedWorkflowActionsLoaderComponent` (see `scoreReview`/`selectSingleReviewer` workflow)
  - `ClaimedTaskActionsLoaderComponent` (see the accept, rejects, ... button in workflow tasks)
  - `ListableObjectComponentLoaderComponent` (check for example item pages/items on search page)
  - `MetadataRepresentationLoaderComponent` (check for example author relation field on publication page)
  - `BrowseBySwitcherComponent` (check the browse pages (both global browse and community & collection browse pages))

## Checklist
- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & specs/tests), or I have provided reasons as to why that's not possible.
- [x] My PR passes [ESLint](https://eslint.org/) validation using `yarn lint`
- [x] My PR doesn't introduce circular dependencies (verified via `yarn check-circ-deps`)
- [x] My PR includes [TypeDoc](https://typedoc.org/) comments for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [x] My PR passes all specs/tests and includes new/updated specs or tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new libraries/dependencies (in `package.json`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR includes new features or configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

tdonohue commented on src/app/shared/abstract-component-loader/abstract-component-loader.component.ts:
```diff
@@ -0,0 +1,136 @@
+  selector: 'ds-abstract-component-loader',
+  templateUrl: './abstract-component-loader.component.html',
+})
+export abstract class AbstractComponentLoaderComponent<T> implements OnInit, OnChanges, OnDestroy {
```
Tiny thing. Can we add a description to this new class?  E.g. how it is used or when it should be extended. That way other developers will understand how to use it for future code.

tdonohue reviewed (approved):
👍 Thanks @alexandrevryghem !  Finally got a chance to test this and look at the code.  I'm not seeing any new bugs based on testing workflow tasks, search, entity object views and all browse by options.  I also looked at the code and it looks good overall (just a minor request for more TypeDoc comments -- see below).

@artlowel : I think this is ready to merge, but it might be worth you taking a glance.  I did want to highlight that this PR appears to be necessary for #2722 .. which itself appears to be necessary to fix the e2e test bugs in the Processes refactoring PRs.   So, the sooner we get this merged, the better.  We have several others dependent on this.

artlowel reviewed (changes requested):
Thanks @alexandrevryghem!

The code looks good and it seems to work well, however, I'm not sure it's a good idea to introduce a lot of abstract component classes that get extended by concrete component classes. Because that way, if you want to make a variant to that abstract super component, _all_ components that extend it also need a variant, even if their code doesn't change. 

The angular team recommends to use injection rather than inheritance for shared code: in other words, move everything that's shared to a service and inject that service in all components that need it. That way, if you want to make a variant, you can simply provide that variant to the existing components instead without further changes.

I concede that the odds of someone needing to make a variant of the `AbstractComponentLoaderComponent` which is the bulk of this PR, are very small. So I'd leave that as is. However here and in #2722, which is built on this, you also add a few other abstract components, such as AbstractBrowseByTypeComponent and ClaimedTaskActionsAbstractComponent. For those, the odds are bigger that someone would want to modify them in a theme for example, so I _would_ refactor those as services.

alexandrevryghem commented:
Thnx @tdonohue & @artlowel for the review!

The latetst push contains these modifications:
- Additional type docs in the `AbstractComponentLoaderComponent`
- I removed the `AbstractBrowseByTypeComponent`
- I removed the `@Input() context: Context;` from the `AbstractBrowseByTypeComponent`, since not all loaders need to have support for multiple contexts. This way it can be customzied for each loader individually (adding support for the context is also simple, you only need to create an `@Input() context: Context` and add 'context' to both the `inputNamesDependentForComponent` & `inputNames` arrays)
- Fixed a bug that caused some dynamically components from being generated twice & ensured that the `ComponentRef` is always destroyed in `OnDestroy` for all the loaders

tdonohue commented:
@artlowel : This is the PR that I mentioned (in today's dev mtg) that we need to move forward in order to unblock #2722 (which builds on this one..and also fixes those e2e test issues you've noted are popping up in other PRs).

Just wanted to highlight the dependencies here... the quicker we can get this PR & #2722 approved & merged, the sooner other PRs can stop hitting that same e2e test failure.

artlowel reviewed (approved):
Thanks @alexandrevryghem!

tdonohue reviewed (approved):
👍 Thanks @alexandrevryghem !  Retested this today...making sure all modified components seem to work the same.  Didn't notice any obvious issues.  So, merging immediately as this is at +2.

(FYI to others: Only *parts* of this effort will be backported to `dspace-7_x`.  See #2800 for details of that backport)
